### PR TITLE
fix(security): defense in depth — input-path cap + cache read-path size check (#95 / #83 follow-up)

### PR DESCRIPTION
**Pull request by sgwannabe**
## Summary

Implements two layers of defense-in-depth around the `idea_summary.maxLength: 5000` schema cap, both deferred from PR #83 to umbrella #95.

- **Layer 1 (input-path)**: `scripts/validate-idea-input.sh` — rejects seed ideas > 5000 Unicode code points BEFORE they reach `runs/<id>/idea.json`, the cache key hasher, or the Socratic system prompt. Wired into `scripts/pre-flight.sh` via new `--idea` / `--idea-file` flags so the existing M1 pre-flight code path actually enforces it (not just documentation). Default mode is **reject**, never silent-truncate. `--truncate` opt-in remains for non-interactive automation.
- **Layer 3 (read-path)**: `scripts/preview-cache.sh::cmd_get` now examines spec-shaped JSON payloads (top-level dict with string `idea_summary`) and reports cache miss (exit 1) if the value exceeds 5000 chars. Belt-and-suspenders: stops a poisoned-on-disk `idea.spec.json` from bypassing the schema gate via a replay-without-revalidate path. Non-spec payloads (e.g. `previews.json` arrays) pass through unchanged.

The schema (`plugins/preview-forge/schemas/idea-spec.schema.json`) remains the canonical authority — neither layer modifies it.

## Layer-1 implementation choice: (b) helper script

Per the task spec, picked option (b) — actual code enforcement via `scripts/validate-idea-input.sh` cited from markdown — rather than (a) prompt-only. Rationale: `commands/new.md` already cites multiple helper scripts (`scripts/preview-cache.sh`, `scripts/detect-surface.sh`, `scripts/recommend-profile.sh`, `scripts/pre-flight.sh`) and uses the W2.7 A-4/A-5 "scripts cited from markdown" pattern. Pure-prompt enforcement would not survive a fresh-run agent that ignores the instruction.

## Side-effect / breaking-change assessment

| # | Concern | Result |
|---|---|---|
| 1 | Existing valid ideas (≤5000 chars) | **No change.** Validator returns rc=0 + echoes input unchanged. `cmd_get` only fires its check when the cached payload is a top-level dict with a string `idea_summary` AND len > 5000 — pure pass-through for current state. |
| 2 | Pre-committed evidence runs `runs/r-clean-20260425-{standard,pro,max}/` | **Safe.** Lengths verified: standard=23, pro=35, max=25 (all far under 5000). |
| 3 | Existing user cache entries `~/.preview-forge/cache/preview-dd/` | **Backwards-compat OK.** Any clean entry hits as before; a hypothetically-poisoned entry forces a regenerate (no data loss — same UX as TTL expiry / Socratic spec change). |
| 4 | `/pf:new` UX change for huge ideas | **Acceptable per S-3 hardening philosophy.** New behavior: huge idea → user sees `validate-idea-input.sh: idea length=N exceeds 5000-character cap. The /pf:new seed idea is bounded at 5000 Unicode code points to match the idea-spec schema's idea_summary maxLength. Please shorten the idea, or re-invoke with --truncate to silently trim to the first 5000 chars.` instead of silent inflate-then-fail-late. |
| 5 | T-7 mock harness `tests/e2e/mock-bootstrap.sh` | **All 3 profiles still pass** (standard / pro / max each report all artifacts present). The harness writes a small canned `idea.spec.json`; never exercises pre-flight directly. |
| 6 | CI surface | **New test sections added** to `tests/fixtures/security/verify-security.sh`: `[#95-L1]` (validator standalone), `[#95-L1-wire]` (pre-flight integration), `[#95-L3]` (cache read-path). Each verifies negative (5001 reject), positive (5000 boundary accept), and edge cases (Unicode codepoint counting, array passthrough, argv/stdin parity, --truncate). |

## Regression suite results (all PASS)

- `bash scripts/verify-plugin.sh` → **57 Pass / 0 Fail**
- `bash tests/fixtures/security/verify-security.sh` → all green incl. 3 new sections
- `bash tests/e2e/mock-bootstrap.sh {standard,pro,max}` → all 3 profiles PASS with expected iframe counts (9 / 18 / 26)
- `bash tests/test-advocate-boilerplate.sh` → 26 advocates share boilerplate hash
- `bash tests/fixtures/lesson07-regression/verify-lesson07.sh` → anchor invariant + panel-bias guards green
- `bash tests/fixtures/cache-concurrency/test-race-window.sh` → alias-first invariant holds
- `bash tests/fixtures/cache-concurrency/test-self-heal.sh` → all 4 self-heal cases OK
- `bash tests/fixtures/cache-concurrency/test-5way.sh` → 5-way concurrent put OK

## Codex review

ONE pass, applied **P1 only** per the workflow constraint:

- **P1 (applied)** — codex flagged that the layer-1 validator was only documented in `commands/new.md`, not wired into `scripts/pre-flight.sh` or `run-supervisor.md`. **Fixed**: added `--idea` / `--idea-file` flags to `pre-flight.sh` (with `[0]` size-cap section that emits `fail` → exit 1), and step 11 in `run-supervisor.md` now instructs the M1 supervisor prompt to invoke the validator before §0.4/§0.8/§0.9. New `[#95-L1-wire]` test section proves the wire-up is enforced (not just documented).
- **P2 (deferred to follow-up)** — codex argued the layer-3 `cmd_get` check guards `idea.spec.json`-shaped payloads but the documented `cmd_put` callers store `previews.json` (top-level array). True for current callers; the layer-3 check is intentionally narrow (per task spec: *"If the cached payload is `idea.spec.json`-shaped..."*) and includes a non-spec-shape passthrough so it does not regress current behavior. The broader question of whether weak-replay of `idea.spec.json` (which is read directly from `runs/<id>/idea.spec.json`, not from `~/.preview-forge/cache/`) deserves its own gate is **out of scope** for this defense-in-depth PR and belongs to a separate hardening pass — not auto-closed here.

## Scope

Closes nothing. This is umbrella **#95** follow-up; other items in that umbrella are still pending and assigned to other agents.

Refs #95, #83.

## 릴리스 노트

* **New Features**
  * Seed idea 입력에 대한 크기 제한 추가 (5000 Unicode code point 이하)
  * Pre-flight 단계에서 검증 기능 강화

* **Bug Fixes**
  * 크기 초과 시 자동 truncation 방지 및 명확한 오류 메시지 제공

* **Tests**
  * 입력 검증 및 캐시 읽기 경로에 대한 보안 회귀 테스트 추가

### CODE REVIEW

File: plugins/preview-forge/agents/meta/run-supervisor.md [hunk 1]
```diff
@@ -47,8 +47,9 @@ model: opus
    - `action == "hint"`: prompt 생략, `/pf:status` 출력에 "💡 Consider --profile=pro next time"로 정적 힌트만
    - `action == "none"`: no-op
 10. **Blackboard 초기화**: `runs/r-<ts>/blackboard.db` 생성 + 초기 row: `(run.pre_flight_passed, ts, cwd, cli_ver, profile, surface, escalation_action)`.
+11. **Idea-input size cap** (umbrella #95 follow-up, deferred from PR #83 — defense-in-depth layer 1): supervisor MUST run `scripts/validate-idea-input.sh - <<< "<idea>"` (or `scripts/pre-flight.sh --idea "<idea>"`) on the raw seed text **BEFORE** §0.4 (`runs/<id>/idea.json` write), §0.8 (`scripts/detect-surface.sh`), §0.9 (`scripts/recommend-profile.sh`), and any cache-key hashing (§4 weak-key probe / §6 strong-key lookup in `commands/new.md`). Validator exits 0 if `len ≤ 5000` Unicode code points, exits 2 otherwise. On exit 2, abort the run and surface the validator's stderr to the user — never silently truncate. The schema's `idea_summary.maxLength: 5000` is the canonical authority; this gate is belt-and-suspenders so a multi-megabyte seed idea cannot inflate the Socratic system prompt or sha256 keyspace before validation fires.
```
[the remark this answers is not in the document]
Comment (sgwannabe) [reply plugins/preview-forge/agents/meta/run-supervisor.md]: here-string(`<<<`) trailing newline 지적 반영했습니다. `run-supervisor.md` §11에서 here-string 예시를 제거하고 positional-arg 형태(`scripts/validate-idea-input.sh "<idea>"`) 또는 `printf '%s' | … -` 파이프 형태로 교체했습니다. here-string이 정확히 5000 code point 시드를 false reject 시킬 수 있다는 점을 본문에 명시하고, `<idea>/<id>/<ts>` placeholder 치환 규약도 함께 문서화했습니다. (commit 6c47b39)

File: scripts/pre-flight.sh [hunk 2]
```diff
@@ -13,6 +13,44 @@ set -euo pipefail
...
+    rm -f "$IDEA_TEXT_FILE"
+  fi
+}
+trap cleanup_idea_tmp EXIT
+
+if [[ "${1:-}" == "--idea" && -n "${2:-}" ]]; then
+  IDEA_TEXT_FILE="$(mktemp -t pf-preflight-idea-XXXXXX)"
+  IDEA_TEXT_FILE_OWNED=1
+  printf '%s' "$2" > "$IDEA_TEXT_FILE"
+  shift 2
```
[the remark this answers is not in the document]
Comment (sgwannabe) [reply scripts/pre-flight.sh]: `-n "${2:-}"` 가 빈 문자열을 silent skip 시킨다는 지적 정확합니다 (codex #6과 동일 이슈). `--idea` 분기를 emptiness와 무관하게 진입하도록 변경하여 `--idea ""` 호출이 `validate-idea-input.sh` 의 empty-reject 분기(T-9.1 parallel)로 hard-fail 하게 만들었습니다. `--idea-file` 도 동일 패턴으로 normalize. tmp file cleanup은 기존 `trap cleanup_idea_tmp EXIT` (line 33-38) 으로 이미 보장되므로 재진입은 생략했습니다 — `set -e` 하에서도 EXIT trap은 발화합니다. (commit 6c47b39)

File: scripts/pre-flight.sh [hunk 3]
```diff
@@ -51,6 +89,28 @@ ok()   { echo "  ✓ $1"; }
...
+  if [[ ! -x "$SCRIPT_DIR/validate-idea-input.sh" ]]; then
+    fail "validate-idea-input.sh missing or not executable at $SCRIPT_DIR/validate-idea-input.sh"
+  else
+    if "$SCRIPT_DIR/validate-idea-input.sh" - < "$IDEA_TEXT_FILE" >/dev/null 2>&1; then
+      ok "idea text within 5000-code-point cap"
+    else
+      validator_rc=$?
+      validator_err=$("$SCRIPT_DIR/validate-idea-input.sh" - < "$IDEA_TEXT_FILE" 2>&1 >/dev/null || true)
+      fail "idea text rejected by validate-idea-input.sh (rc=$validator_rc): ${validator_err}"
+    fi
```
[the remark this answers is not in the document]
Comment (sgwannabe) [reply scripts/pre-flight.sh]: validator double-execution 비효율 반영했습니다. `validator_err=$("$SCRIPT_DIR/validate-idea-input.sh" - < "$IDEA_TEXT_FILE" 2>&1 >/dev/null)` 한 번 호출로 rc(if 분기)와 stderr message를 동시에 캡처하도록 변경 — `2>&1 >/dev/null` 순서로 stderr→stdout 리다이렉트 후 stdout만 버려서 $(…)는 error message만 수집합니다. 실패 경로에서 python3 호출이 절반으로 줄어듭니다. 후행 newline 이슈는 stderr의 마지막 줄 trailing newline만 영향이고 `fail "…: ${validator_err}"` 출력에 의도적으로 압축되는 게 적절하다고 판단했습니다. (commit 6c47b39)

File: scripts/validate-idea-input.sh [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+fi
+
+# Length check via python3 (Unicode code points, matching JSON Schema
+# `maxLength` semantics — see header). Argv pass + with-open + single-
+# quoted heredoc closes the inline-string interpolation surface (same
+# pattern as scripts/preview-cache.sh::py_read_json caller contract).
+length=$(printf '%s' "$idea" | python3 -c '
+import sys
+print(len(sys.stdin.read()))
+')
```
[the remark this answers is not in the document]
Comment (sgwannabe) [reply scripts/validate-idea-input.sh]: bounded read 지적 반영했습니다. `sys.stdin.read(MAX_LEN+1)` 로 5001 code points 까지만 메모리에 적재하여 5MB+ 시드도 peak RSS 가 일정합니다. `MAX_LEN` 은 single-quoted python source 외부에서 argv(`sys.argv[1]`)로 전달하여 shell-interpolation surface 도 같이 닫았습니다 (gemini #5와 동일 패턴). 추가 보강: `set -euo pipefail` 환경에서 python이 일찍 종료하면 upstream `printf` 가 SIGPIPE→rc=141 로 죽어 pipefail이 전체를 실패시키므로, python 측에서 chunked drain loop(`while sys.stdin.read(1<<20):`)을 돌려 SIGPIPE 를 회피했습니다. 1M / 5M payload 모두 rc=2 로 정상 reject 확인. (commit 6c47b39)

File: scripts/validate-idea-input.sh [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+fi
+
+# Over the cap.
+if [[ "$mode" == "truncate" ]]; then
+  echo "validate-idea-input.sh: idea length=$length > $MAX_LEN — truncating to first $MAX_LEN code points" >&2
+  printf '%s' "$idea" | python3 -c "
+import sys
+data = sys.stdin.read()
+sys.stdout.write(data[:$MAX_LEN])
+"
```
[the remark this answers is not in the document]
Comment (sgwannabe) [reply scripts/validate-idea-input.sh]: --truncate 분기에도 동일하게 `sys.stdin.read(MAX_LEN)` bounded read + argv pass 적용했습니다. inline python source 가 single-quoted 로 변환되어 `$MAX_LEN` shell interpolation surface 가 사라졌습니다 — `scripts/preview-cache.sh::py_read_json` 의 caller contract 와 동일한 패턴입니다. truncate branch 도 SIGPIPE 회피용 chunked drain loop 추가. 1M payload 입력 → 정확히 5000 byte 출력 확인. (commit 6c47b39)

File: scripts/pre-flight.sh [hunk 6]
```diff
@@ -13,6 +13,44 @@ set -euo pipefail
...
+IDEA_TEXT_FILE=""
+IDEA_TEXT_FILE_OWNED=0   # 1 = we created it, must rm on exit
+cleanup_idea_tmp() {
+  if [[ "$IDEA_TEXT_FILE_OWNED" -eq 1 && -n "$IDEA_TEXT_FILE" && -f "$IDEA_TEXT_FILE" ]]; then
+    rm -f "$IDEA_TEXT_FILE"
+  fi
+}
+trap cleanup_idea_tmp EXIT
+
+if [[ "${1:-}" == "--idea" && -n "${2:-}" ]]; then
```
[the remark this answers is not in the document]
Comment (sgwannabe) [reply scripts/pre-flight.sh]: P2 정확합니다 — gemini #2 와 동일 이슈로 함께 처리했습니다. `if [[ "${1:-}" == "--idea" ]]; then` 으로 emptiness 무관하게 분기 진입 + `[[ $# -lt 2 ]]` usage error 가드 추가 + `printf '%s' "$2" > $IDEA_TEXT_FILE` 로 빈 문자열도 그대로 tmp 에 쓴 뒤 `validate-idea-input.sh` 의 empty-reject(T-9.1 parallel)로 hard-fail 시킵니다. `--idea-file` 도 동일 패턴 normalize. `bash scripts/pre-flight.sh --idea ""` 실행 시 §0 에서 "idea text is empty — refusing" 으로 rc=1 종료 확인 (verify-security #95-L1-wire 회귀 통과). (commit 6c47b39)
